README.md
# JDK Utility Library

[![NPM Version][npm-image]][npm-url]
[![NPM Downloads][downloads-image]][downloads-url]
[![Travis CI Build][travis-image]][travis-url]
[![Deps][david-image]][david-url]
[![Dev Deps][david-dev-image]][david-dev-url]

> Note: jdklib requires Node.js 8.12.0 or newer.

## Installation

    npm install jdklib

## Examples

Detect installed JDKs:

```javascript
import * as jdklib from 'jdklib';

jdklib
    .detect()
    .then(jdkInfo => {
        console.info(jdkInfo);
    })
    .catch(err => console.error);
```

## Reporting Bugs or Submitting Fixes

If you run into problems, and trust us, there are likely plenty of them at this
point -- please create an [Issue](https://github.com/appcelerator/jdklib/issues)
or, even better, send us a pull request.

## Contributing

jdklib is an open source project. jdklib wouldn't be where it is now without
contributions by the community. Please consider forking jdklib to improve,
enhance or fix issues. If you feel like the community will benefit from your
fork, please open a pull request.

To protect the interests of the jdklib contributors, Appcelerator, customers
and end users we require contributors to sign a Contributors License Agreement
(CLA) before we pull the changes into the main repository. Our CLA is simple and
straightforward - it requires that the contributions you make to any
Appcelerator open source project are properly licensed and that you have the
legal authority to make those changes. This helps us significantly reduce future
legal risk for everyone involved. It is easy, helps everyone, takes only a few
minutes, and only needs to be completed once.

[You can digitally sign the CLA](http://bit.ly/app_cla) online. Please indicate
your email address in your first pull request so that we can make sure that will
locate your CLA.  Once you've submitted it, you no longer need to send one for
subsequent submissions.

## Legal

This project is open source under the [Apache Public License v2][1] and is developed by
[Axway, Inc](http://www.axway.com/) and the community. Please read the [`LICENSE`][1] file included
in this distribution for more information.

[1]: https://github.com/appcelerator/androidlib/blob/master/LICENSE

[npm-image]: https://img.shields.io/npm/v/jdklib.svg
[npm-url]: https://npmjs.org/package/jdklib
[downloads-image]: https://img.shields.io/npm/dm/jdklib.svg
[downloads-url]: https://npmjs.org/package/jdklib
[travis-image]: https://img.shields.io/travis/appcelerator/jdklib.svg
[travis-url]: https://travis-ci.org/appcelerator/jdklib
[david-image]: https://img.shields.io/david/appcelerator/jdklib.svg
[david-url]: https://david-dm.org/appcelerator/jdklib
[david-dev-image]: https://img.shields.io/david/dev/appcelerator/jdklib.svg
[david-dev-url]: https://david-dm.org/appcelerator/jdklib#info=devDependencies


CHANGELOG.md
# v4.0.0 (Dec 1, 2020)

 * BREAKING CHANGE: Bumped minimum required Node.js version from v8.12.0 to v10.13.0.
 * chore: Updated dependencies.

# v3.0.2 (Jan 8, 2020)

 * chore: Updated dependencies.

# v3.0.1 (Nov 7, 2019)

 * chore: Updated dependencies.

# v3.0.0 (Aug 14, 2019)

 * BREAKING CHANGE: Bumped minimum required Node.js version from v8.0.0 to v8.12.0.
 * chore: Updated dependencies.

# v2.4.1 (Jun 5, 2019)

 * chore: Updated dependencies.

# v2.4.0 (Mar 29, 2019)

 * chore: Updated dependencies.

# v2.3.1 (Jan 25, 2019)

 * chore: Updated dependencies.

# v2.3.0 (Jan 18, 2019)

 * chore: Upgraded to Gulp 4.
 * refactor: Refactored promises to async/await.
 * chore: Updated dependencies.

# v2.2.1 (Apr 9, 2018)

 * chore: Updated dependencies.

# v2.2.0 (Dec 8, 2017)

# v2.1.0 (Oct 19, 2017)

# v2.0.4 (Oct 18, 2017)

# v2.0.3 (Oct 17, 2017)

# v2.0.2 (Oct 10, 2017)

# v2.0.1 (Sep 9, 2017)

# v2.0.0 (Aug 28, 2017)


test/test-jdk.js
import JDK, { detect } from '../dist/index';

import path from 'path';
import { exe } from 'appcd-subprocess';

describe('JDK', function () {
	this.timeout(5000);

	it('should throw error if dir is not a string', () => {
		expect(() => {
			new JDK();
		}).to.throw(TypeError, 'Expected directory to be a valid string');

		expect(() => {
			new JDK(123);
		}).to.throw(TypeError, 'Expected directory to be a valid string');
	});

	it('should throw error if dir does not exist', async () => {
		await expect(detect('doesnotexist')).to.be.rejectedWith(Error, 'Directory does not exist');
	});

	it('should error if dir is missing the JVM library', async () => {
		await expect(detect(path.join(__dirname, 'mocks', 'empty'))).to.be.rejectedWith(Error, 'Directory missing JVM library');
	});

	it('should error if dir is missing essential jdk tools', async () => {
		await expect(detect(path.join(__dirname, 'mocks', 'incomplete-jdk'))).to.be.rejectedWith(Error, 'Directory missing required program');
	});

	it('should detect JDK 1.6', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-1.6');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('64bit');
		expect(jdk.build).to.equal(45);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			javac:     path.join(dir, 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.equal('1.6.0');
	});

	it('should detect JDK 1.7', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-1.7');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('64bit');
		expect(jdk.build).to.equal(80);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			javac:     path.join(dir, 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.equal('1.7.0');
	});

	it('should detect JDK 1.8 64-bit', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-1.8');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('64bit');
		expect(jdk.build).to.equal(92);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			javac:     path.join(dir, 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.equal('1.8.0');
	});

	it('should detect JDK 1.8 32-bit', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-1.8-32bit');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('32bit');
		expect(jdk.build).to.equal(92);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			javac:     path.join(dir, 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.equal('1.8.0');
	});

	(process.platform === 'darwin' ? it : it.skip)('should detect JDK 1.8 64-bit with macOS pathing', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-1.8-darwin');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('64bit');
		expect(jdk.build).to.equal(92);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'Contents', 'Home', 'bin', 'java' + exe),
			javac:     path.join(dir, 'Contents', 'Home', 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'Contents', 'Home', 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'Contents', 'Home', 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(path.join(dir, 'Contents', 'Home'));
		expect(jdk.version).to.equal('1.8.0');
	});

	it('should detect JDK 9', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-9');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('64bit');
		expect(jdk.build).to.equal(181);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			javac:     path.join(dir, 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.equal('9');
	});

	(process.platform === 'darwin' ? it : it.skip)('should detect JDK 9 with macOS pathing', async () => {
		const dir = path.join(__dirname, 'mocks', 'jdk-9-darwin');
		const jdk = await detect(dir);

		expect(jdk.arch).to.equal('64bit');
		expect(jdk.build).to.equal(11);
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'Contents', 'Home', 'bin', 'java' + exe),
			javac:     path.join(dir, 'Contents', 'Home', 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'Contents', 'Home', 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'Contents', 'Home', 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(path.join(dir, 'Contents', 'Home'));
		expect(jdk.version).to.equal('9.0.1');
	});

	it('should not detect version or arch if javac is not found', async () => {
		const dir = path.join(__dirname, 'mocks', 'bad-bin-jdk');
		let jdk = new JDK(dir);
		delete jdk.executables.javac;
		jdk = await jdk.init();

		expect(jdk.arch).to.be.null;
		expect(jdk.build).to.be.null;
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.be.null;
	});

	it('should not detect version if javac is bad', async () => {
		const dir = path.join(__dirname, 'mocks', 'bad-bin-jdk');
		const jdk = await detect(dir);

		expect(jdk.arch).to.be.null;
		expect(jdk.build).to.be.null;
		expect(jdk.executables).to.deep.equal({
			java:      path.join(dir, 'bin', 'java' + exe),
			javac:     path.join(dir, 'bin', 'javac' + exe),
			keytool:   path.join(dir, 'bin', 'keytool' + exe),
			jarsigner: path.join(dir, 'bin', 'jarsigner' + exe)
		});
		expect(jdk.path).to.equal(dir);
		expect(jdk.version).to.be.null;
	});
});


gulpfile.js
'use strict';

const $ = require('gulp-load-plugins')();
const del = require('del');
const gulp = require('gulp');
const manifest = require('./package.json');
const path = require('path');

const coverageDir = path.join(__dirname, 'coverage');
const distDir = path.join(__dirname, 'dist');
const docsDir = path.join(__dirname, 'docs');

/*
 * Clean tasks
 */
gulp.task('clean', ['clean-coverage', 'clean-dist', 'clean-docs']);

gulp.task('clean-coverage', done => { del([coverageDir]).then(() => done()) });

gulp.task('clean-dist', done => { del([distDir]).then(() => done()) });

gulp.task('clean-docs', done => { del([docsDir]).then(() => done()) });

/*
 * build tasks
 */
gulp.task('build', ['clean-dist', 'lint-src'], () => {
	return gulp
		.src('src/**/*.js')
		.pipe($.plumber())
		.pipe($.debug({ title: 'build' }))
		.pipe($.sourcemaps.init())
		.pipe($.babel())
		.pipe($.sourcemaps.write('.'))
		.pipe(gulp.dest(distDir));
});

gulp.task('docs', ['lint-src', 'clean-docs'], () => {
	return gulp.src('src')
		.pipe($.plumber())
		.pipe($.debug({ title: 'docs' }))
		.pipe($.esdoc({
			// debug: true,
			destination: docsDir,
			plugins: [
				{ name: 'esdoc-es7-plugin' }
			],
			title: manifest.name
		}));
});

/*
 * lint tasks
 */
function lint(pattern) {
	return gulp.src(pattern)
		.pipe($.plumber())
		.pipe($.eslint())
		.pipe($.eslint.format())
		.pipe($.eslint.failAfterError());
}

gulp.task('lint-src', () => lint('src/**/*.js'));

gulp.task('lint-test', () => lint('test/**/test-*.js'));

/*
 * test tasks
 */
gulp.task('test', ['lint-src', 'lint-test'], () => {
	let suite;
	let grep;
	let p = process.argv.indexOf('--suite');
	if (p !== -1 && p + 1 < process.argv.length) {
		suite = process.argv[p + 1];
	}
	p = process.argv.indexOf('--grep');
	if (p !== -1 && p + 1 < process.argv.length) {
		grep = process.argv[p + 1];
	}

	return gulp.src(['src/**/*.js', 'test/**/*.js'])
		.pipe($.plumber())
		.pipe($.debug({ title: 'build' }))
		.pipe($.sourcemaps.init())
		.pipe($.babel())
		.pipe($.sourcemaps.write())
		.pipe($.injectModules())
		.pipe($.filter(suite ? ['test/setup.js'].concat(suite.split(',').map(s => 'test/**/test-' + s + '.js')) : 'test/**/*.js'))
		.pipe($.debug({ title: 'test' }))
		.pipe($.mocha({ grep: grep }));
});

gulp.task('coverage', ['lint-src', 'lint-test', 'clean-coverage'], function (cb) {
	gulp.src('src/**/*.js')
		.pipe($.plumber())
		.pipe($.debug({ title: 'build' }))
		.pipe($.sourcemaps.init())
		.pipe($.babelIstanbul())
		.pipe($.sourcemaps.write())
		.pipe($.injectModules())
		.on('finish', function () {
			gulp.src('test/**/*.js')
				.pipe($.plumber())
				.pipe($.debug({ title: 'test' }))
				.pipe($.sourcemaps.init())
				.pipe($.babel())
				.pipe($.sourcemaps.write())
				.pipe($.injectModules())
				.pipe($.mocha())
				.pipe($.babelIstanbul.writeReports())
				.on('end', cb);
		});
});

gulp.task('default', ['build']);


src/index.js
/* istanbul ignore if */
if (!Error.prepareStackTrace) {
	require('source-map-support/register');
}

import path from 'path';
import appcdLogger from 'appcd-logger';

import { isDir, isFile } from 'appcd-fs';
import { expandPath, real } from 'appcd-path';
import { exe, run } from 'appcd-subprocess';

const { log } = appcdLogger('jdklib');
const { highlight } = appcdLogger.styles;

const re = /^javac (.+?)(?:_(.+))?$/;

/**
 * Common JDK install locations.
 * @type {Object}
 */
export const jdkLocations = {
	darwin: [
		'/Library/Java/JavaVirtualMachines',
		'/System/Library/Java/JavaVirtualMachines'
	],
	linux: [
		'/usr/lib/jvm'
	]
	// note: for Windows, we check the Windows Registry
};

/**
 * Common search paths for the JVM library. This is used only for validating if
 * a directory is a JDK.
 * @type {Object}
 */
export const libjvmLocations = {
	linux: [
		'lib/amd64/client/libjvm.so',
		'lib/amd64/server/libjvm.so',
		'lib/i386/client/libjvm.so',
		'lib/i386/server/libjvm.so',
		'jre/lib/amd64/client/libjvm.so',
		'jre/lib/amd64/server/libjvm.so',
		'jre/lib/i386/client/libjvm.so',
		'jre/lib/i386/server/libjvm.so',
		'lib/server/libjvm.so'
	],
	darwin: [
		'jre/lib/server/libjvm.dylib',
		'../Libraries/libjvm.dylib',
		'lib/server/libjvm.dylib'
	],
	win32: [
		'jre/bin/server/jvm.dll',
		'jre/bin/client/jvm.dll',
		'bin/server/jvm.dll'
	]
};

/**
 * JDK information object.
 */
export default class JDK {
	/**
	 * Performs some simple tests to check if the specified directory is a JDK or not, then
	 * initializes the JDK info.
	 *
	 * @param {String} dir - The directory to scan.
	 * @access public
	 */
	constructor(dir) {
		if (typeof dir !== 'string' || !dir) {
			throw new TypeError('Expected directory to be a valid string');
		}

		dir = expandPath(dir);
		if (!isDir(dir)) {
			throw new Error('Directory does not exist');
		}

		// on OS X, the JDK lives in Contents/Home
		if (process.platform === 'darwin') {
			const p = path.join(dir, 'Contents', 'Home');
			if (isDir(p)) {
				dir = p;
			}
		}

		const libjvms = libjvmLocations[process.platform];
		if (!libjvms || !libjvms.some(p => isFile(path.resolve(dir, p)))) {
			throw new Error('Directory missing JVM library');
		}

		this.arch        = null;
		this.build       = null;
		this.executables = {};
		this.path        = dir;
		this.version     = null;

		if (![ 'java', 'javac', 'keytool', 'jarsigner' ].every(cmd => {
			const p = path.join(dir, 'bin', cmd + exe);
			if (isFile(p)) {
				this.executables[cmd] = real(p);
				return true;
			}
			return false;
		})) {
			throw new Error('Directory missing required program');
		}

		log('Found a JDK, but need to init: %s', highlight(dir));
	}

	/**
	 * Fetches the JDK version and architecture by running javac.
	 *
	 * @returns {Promise}
	 * @access public
	 */
	async init() {
		const javac = this.executables.javac;
		if (!javac) {
			log('No javac found, skipping version detection');
			return Promise.resolve(this);
		}

		// we do a nested try/catch because it's ok if this fails since our goal is to detect, not
		// perform some task
		try {
			let stderr, stdout, arch;

			try {
				// try the 64-bit version first
				({ stderr, stdout } = await run(javac, [ '-d64', '-version' ]));
				log('javac is the 64-bit version');
				arch = '64bit';
			} catch (err) {
				// if err.code === 2, then we have the 64-bit version, but we must re-run javac to
				// get the version since on Windows it doesn't print the version correctly
				({ stdout, stderr } = await run(javac, [ '-version' ]));
				log(`javac is the ${err.code === 2 ? 64 : 32}-bit version`);
				arch = err.code === 2 ? '64bit' : '32bit';
			}

			const m = stderr.trim().match(re) || stdout.trim().match(re);
			let build = m && parseInt(m[2]);
			if (!build) {
				const { stderr } = await run(this.executables.java, [ '-version' ]);
				const m = stderr.trim().match(/\(build .+?\+(\d+)\)/);
				build = m && parseInt(m[1]);
			}
			this.version = m && m[1] || null;
			this.build = build;
			this.arch = arch;
		} catch (err) {
			// squelch
		}

		return this;
	}
}

/**
 * Detects if the specified directory contains a JDK and if so, returns a `JDK` object.
 *
 * @param {String} dir - The directory to scan.
 * @returns {Promise}
 */
export async function detect(dir) {
	return await new JDK(dir).init();
}
